### PR TITLE
Add RM68090 driver and TFT identification code based on MCUFRIEND_kbv

### PR DESCRIPTION
**Pull request by NoNamedCat**

### CODE REVIEW

File: drivers/video/fbdev/r61520fb.c [hunk 1]
```diff
@@ -396,10 +396,81 @@ static irqreturn_t lcdc_irq_handler(int irq, void *arg)
...
+  suniv_clrbits(iomm.lcdc + PE_DATA, (11 << 0));
+  mdelay(150);
+  suniv_setbits(iomm.lcdc + PE_DATA, (11 << 0));
+  mdelay(150);
+  
+  printk(" ");
+  printk("TFT IDENTIFICATION 8BITS:");
+  printk(" ");
+  
+    readReg(0x00, 2, "ID: ILI9320, ILI9325, ILI9335, ...");
```
[the remark this answers is not in the document]
Comment (nfriedly) [reply drivers/video/fbdev/r61520fb.c:432]:
```suggestion
    readReg(0x00, 2, "ID"); // ILI9320, ILI9325, ILI9335, ...
```

File: drivers/video/fbdev/rm68090fb.c [hunk 2]
```diff
@@ -0,0 +1,1124 @@
...
+// Should be a number 1-4, but I'm not yet sure which numbers correspond to which panels
+// 1 = R61520 - R61520 doesn't support the 0x04 read_DDB_start / Read Display ID Information command.
+// 2 = ST7789S
+// 3 = R61505W
+// 4 = ?
+// Will be automatically detected if left unset (which usually works)
+static uint32_t version=0;
+static bool invert=false;
+module_param(version,uint,0660);
+module_param(invert,bool,0660);
```
Comment (nfriedly) [inline drivers/video/fbdev/rm68090fb.c:80]: We probably don't need the version stuff now that we have separate drivers for separate panels. Although it's not going to be as easy as just deleting it; there's some code below that still checks both this and the detected version. Doesn't necessarily need to be changed in this PR, but it wouldn't hurt.

File: drivers/video/fbdev/rm68090fb.c [hunk 3]
```diff
@@ -0,0 +1,1124 @@
...
+
+static uint32_t lcdc_rd_dat(void)
+{
+	while(lcdc_wait_busy());
+	suniv_setbits(iomm.lcdc + TCON0_CPU_IF_REG, (1 << 25)); // ca=1
+	while(lcdc_wait_busy());
+	return extend_24b_to_16b(readl(iomm.lcdc + TCON0_CPU_RD_REG));
+}
+
+static void gpio_wr(uint32_t is_data, uint32_t val)
```
Comment (tiopex) [inline drivers/video/fbdev/rm68090fb.c:230]: not needed to remove

File: drivers/video/fbdev/rm68090fb.c [hunk 4]
```diff
@@ -0,0 +1,1124 @@
...
+  ret = (val & 0x00ff) << 1;
+  ret|= (val & 0xff00) << 2;
+  ret|= is_data ? 0x80000 : 0;
+  ret|= 0x100000; 
+  writel(ret, iomm.gpio + PD_DATA);
+  ret|= 0x40000;
+  writel(ret, iomm.gpio + PD_DATA);
+}
+
+static void ser_wr(uint8_t is_data, uint8_t data)
```
Comment (tiopex) [inline drivers/video/fbdev/rm68090fb.c:243]: not needed to remove

File: drivers/video/fbdev/rm68090fb.c [hunk 5]
```diff
@@ -0,0 +1,1124 @@
...
+		else{
+			suniv_clrbits(iomm.gpio + PE_DATA, 1 << 1); // sda
+		}
+    udelay(10);
+  	suniv_setbits(iomm.gpio + PC_DATA, 1 << 3); // dcx
+    udelay(10);
+  }
+}
+
+static void ser_wr_cmd(uint8_t data)
```
Comment (tiopex) [inline drivers/video/fbdev/rm68090fb.c:268]: not needed to remove

File: drivers/video/fbdev/rm68090fb.c [hunk 6]
```diff
@@ -0,0 +1,1124 @@
...
+{
+  ser_wr(0, data);
+}
+ 
+static void ser_wr_dat(uint8_t data)
```
Comment (tiopex) [inline drivers/video/fbdev/rm68090fb.c:273]: not needed to remove

File: drivers/video/fbdev/rm68090fb.c [hunk 7]
```diff
@@ -0,0 +1,1124 @@
...
+{
+  ser_wr(1, data);
+}
+
+static uint32_t ser_rd_dat(void)
```
Comment (tiopex) [inline drivers/video/fbdev/rm68090fb.c:278]: not needed to remove

File: drivers/video/fbdev/rm68090fb.c [hunk 8]
```diff
@@ -0,0 +1,1124 @@
...
+			ret|= 1;
+		}
+    udelay(10);
+  	suniv_setbits(iomm.gpio + PC_DATA, 1 << 3); // dcx
+  }
+	suniv_setbits(iomm.gpio + PE_CFG0, 1 << 4); // sda
+	return ret;
+}
+
+static void ser_init(void)
```
Comment (tiopex) [inline drivers/video/fbdev/rm68090fb.c:300]: not needed to remove

File: drivers/video/fbdev/rm68090fb.c [hunk 9]
```diff
@@ -0,0 +1,1124 @@
...
+
+  suniv_setbits(iomm.gpio + PE_CFG0, 1 << 4); // sda
+  suniv_clrbits(iomm.gpio + PE_CFG0, 1 << 5);
+  suniv_clrbits(iomm.gpio + PE_CFG0, 1 << 6);
+  suniv_setbits(iomm.gpio + PE_DATA, 1 << 1);
+  suniv_setbits(iomm.gpio + PE_PUL0, 1 << 2);
+  suniv_clrbits(iomm.gpio + PE_PUL0, 1 << 3);
+}
+
+static void ser_deinit(void)
```
Comment (tiopex) [inline drivers/video/fbdev/rm68090fb.c:327]: not needed to remove

File: drivers/video/fbdev/rm68090fb.c [hunk 10]
```diff
@@ -0,0 +1,1124 @@
...
+  suniv_clrbits(iomm.gpio + PE_CFG0, 1 << 0);
+  suniv_clrbits(iomm.gpio + PE_CFG0, 1 << 1);
+  suniv_clrbits(iomm.gpio + PE_CFG0, 1 << 2);
+
+  suniv_clrbits(iomm.gpio + PE_CFG0, 1 << 4);
+  suniv_clrbits(iomm.gpio + PE_CFG0, 1 << 5);
+  suniv_clrbits(iomm.gpio + PE_CFG0, 1 << 6);
+}
+
+static void gpio_wr_cmd(uint32_t val)
```
Comment (tiopex) [inline drivers/video/fbdev/rm68090fb.c:338]: not needed to remove

File: drivers/video/fbdev/rm68090fb.c [hunk 11]
```diff
@@ -0,0 +1,1124 @@
...
+{
+  gpio_wr(0, val);
+}
+
+static void gpio_wr_dat(uint32_t val)
```
Comment (tiopex) [inline drivers/video/fbdev/rm68090fb.c:343]: not needed to remove

File: drivers/video/fbdev/rm68090fb.c [hunk 12]
```diff
@@ -0,0 +1,1124 @@
...
+{
+  gpio_wr(1, val);
+}
+
+static void refresh_lcd(struct myfb_par *par)
+{
+  if((miyoo_ver <= 2) || (miyoo_ver == 4)){
```
Comment (tiopex) [inline drivers/video/fbdev/rm68090fb.c:350]: this if is not needed - to remove

File: drivers/video/fbdev/rm68090fb.c [hunk 13]
```diff
@@ -0,0 +1,1124 @@
...
+      }
+      refresh_lcd(par);
+    }
+  }
+  return IRQ_HANDLED;
+}
+
+static int panel_init(void)
+{
+  uint16_t x, ver[4]={0};
```
Comment (tiopex) [inline drivers/video/fbdev/rm68090fb.c:406]: not needed

File: drivers/video/fbdev/rm68090fb.c [hunk 14]
```diff
@@ -0,0 +1,1124 @@
...
+  ser_wr_cmd(0x04);
+  if(ser_rd_dat() == 0x42c2a97f){
+    ver = 4;
+  }
+  ser_deinit();
+
+  if (version == 4 && ver != 4) {
+    printk("Warning: LCD controller not detected as version 4, but being overridden by module_param as version 4 (detection 1 of 2)");
+    ver = version;
+  }
```
Comment (tiopex) [inline drivers/video/fbdev/rm68090fb.c:574]: ser_init();
  ser_wr_cmd(0x04);
  if(ser_rd_dat() == 0x42c2a97f){
    ver = 4;
  }
  ser_deinit();

  if (version == 4 && ver != 4) {
    printk("Warning: LCD controller not detected as version 4, but being overridden by module_param as version 4 (detection 1 of 2)");
    ver = version;
  }

to remove

File: drivers/video/fbdev/rm68090fb.c [hunk 15]
```diff
@@ -0,0 +1,1124 @@
...
+  writel((5 << 8), iomm.debe + DEBE_LAY1_ATT_CTRL_REG1);
+	suniv_setbits(iomm.debe + DEBE_MODE_CTRL_REG, (1 << 8));
+	suniv_setbits(iomm.debe + DEBE_REGBUFF_CTRL_REG, (1 << 1));
+  suniv_setbits(iomm.debe + DEBE_MODE_CTRL_REG, (1 << 1));
+
+	ret = readl(iomm.lcdc + TCON_CTRL_REG);
+	ret&= ~(1 << 0);
+	writel(ret, iomm.lcdc + TCON_CTRL_REG);
+	ret = (v_front_porch + v_back_porch + v_sync_len);
+  if(ver == 4){
```
Comment (tiopex) [inline drivers/video/fbdev/rm68090fb.c:600]: whole content of if(ver == 4){ to remove

File: drivers/video/fbdev/r61520fb.c [hunk 16]
```diff
@@ -396,10 +396,80 @@ static irqreturn_t lcdc_irq_handler(int irq, void *arg)
...
+    readReg(0xD4, 4, "Novatek");
+    readReg(0xDA, 2, "RDID1");
+    readReg(0xDB, 2, "RDID2");
+    readReg(0xDC, 2, "RDID3");
+    readReg(0xE0, 16, "GAMMA-P");
+    readReg(0xE1, 16, "GAMMA-N");
+    readReg(0xEF, 6, "ILI9327");
+    readReg(0xF2, 12, "Adjust Control 2");
+    readReg(0xF6, 4, "Interface Control");
+
```
Comment (tiopex) [inline drivers/video/fbdev/r61520fb.c:471]: I suggest adding this code under some if. e.g if (debugEnabled == true ) {

File: drivers/video/fbdev/Kconfig [hunk 17]
```diff
@@ -2532,4 +2532,12 @@ config FB_ST7789S_TE
 	select FB_SYS_IMAGEBLIT
 	help
 	  Framebuffer driver for ST7789S SLCD panel.
-	  
\ No newline at end of file
+
+config RM68090
+	tristate "RM68090 SLCD framebuffer support with TE pin"
```
Comment (tiopex) [inline drivers/video/fbdev/Kconfig:2537]: tristate "RM68090 SLCD framebuffer support"

remove "with TE pin"

File: drivers/video/fbdev/r61520fb.c [hunk 18]
```diff
@@ -396,10 +402,81 @@ static irqreturn_t lcdc_irq_handler(int irq, void *arg)
...
+  suniv_setbits(iomm.lcdc + PE_DATA, (11 << 0));
+  mdelay(50);
+  suniv_clrbits(iomm.lcdc + PE_DATA, (11 << 0));
+  mdelay(150);
+  suniv_setbits(iomm.lcdc + PE_DATA, (11 << 0));
+  mdelay(150);
+  
+  printk(" ");
+  printk("TFT IDENTIFICATION 8BITS:");
+  printk(" ");
```
Comment (tiopex) [inline drivers/video/fbdev/r61520fb.c:435]: printk(" ");
  printk("TFT IDENTIFICATION 8BITS:");
  printk(" ");

should be also under debug

File: drivers/video/fbdev/Kconfig (hunk 17 above)
Comment (NoNamedCat) [reply drivers/video/fbdev/Kconfig:2537]: Done

Comment (tiopex) [reply drivers/video/fbdev/Kconfig:2537]: remove "with TE pin" word

File: drivers/video/fbdev/rm68090fb.c [hunk 19]
```diff
@@ -0,0 +1,1124 @@
...
+static bool flip=false;
+module_param(flip,bool,0660);
+
+// Which LCD controller are we driving? E.g. R61520, R61526, R61581, etc. 
+// Should be a number 1-4, but I'm not yet sure which numbers correspond to which panels
+// 1 = R61520 - R61520 doesn't support the 0x04 read_DDB_start / Read Display ID Information command.
+// 2 = ST7789S
+// 3 = R61505W
+// 4 = ?
+// Will be automatically detected if left unset (which usually works)
```
Comment (tiopex) [inline drivers/video/fbdev/rm68090fb.c:76]: these comments are not valid here

**Comment by TriForceX** [comment]
Closed due this https://github.com/MiyooCFW/kernel/pull/32

**Pull request closed without merging**